x-pack/plugin/ml-feature-index-builder/src/main/java/org/elasticsearch/xpack/ml/featureindexbuilder/job/FeatureIndexBuilderJobConfig.java

```diff
@@ -28,25 +30,55 @@
+    private static final ParseField INDEX_PATTERN = new ParseField("index_pattern");
+    private static final ParseField DESTINATION_INDEX = new ParseField("destination_index");
+    private static final ParseField SOURCES = new ParseField("sources");
+    private static final ParseField AGGREGATIONS = new ParseField("aggregations");
 
     private final String id;
+    private final String indexPattern;
+    private final String destinationIndex;
+    private final SourceConfig sourceConfig;
+    private final AggregationConfig aggregationConfig;
 
     private static final ConstructingObjectParser<FeatureIndexBuilderJobConfig, String> PARSER = new ConstructingObjectParser<>(NAME, false,
             (args, optionalId) -> {
                 String id = args[0] != null ? (String) args[0] : optionalId;
-                return new FeatureIndexBuilderJobConfig(id);
+                String indexPattern = (String) args[1];
+                String destinationIndex = (String) args[2];
+                SourceConfig sourceConfig= (SourceConfig) args[3];
+                AggregationConfig aggregationConfig = (AggregationConfig) args[4];
+                return new FeatureIndexBuilderJobConfig(id, indexPattern, destinationIndex, sourceConfig, aggregationConfig);
             });
 
     static {
         PARSER.declareString(optionalConstructorArg(), ID);
+        PARSER.declareString(constructorArg(), INDEX_PATTERN);
+        PARSER.declareString(constructorArg(), DESTINATION_INDEX);
+        PARSER.declareObject(optionalConstructorArg(), (p, c) -> SourceConfig.fromXContent(p), SOURCES);
+        PARSER.declareObject(optionalConstructorArg(), (p, c) -> AggregationConfig.fromXContent(p), AGGREGATIONS);
     }
 
-    public FeatureIndexBuilderJobConfig(final String id) {
-        this.id = id;
+    public FeatureIndexBuilderJobConfig(final String id,
+                                        final String indexPattern,
+                                        final String destinationIndex,
+                                        final SourceConfig sourceConfig,
+                                        final AggregationConfig aggregationConfig) {
+        this.id = ExceptionsHelper.requireNonNull(id, ID.getPreferredName());
+        this.indexPattern = ExceptionsHelper.requireNonNull(indexPattern, INDEX_PATTERN.getPreferredName());
+        this.destinationIndex = ExceptionsHelper.requireNonNull(destinationIndex, DESTINATION_INDEX.getPreferredName());
+
+        // TODO: check for null?
+        this.sourceConfig = sourceConfig;
+        this.aggregationConfig = aggregationConfig;
     }
 
     public FeatureIndexBuilderJobConfig(final StreamInput in) throws IOException {
         id = in.readString();
+        indexPattern = in.readString();
+        destinationIndex = in.readString();
+        sourceConfig = in.readOptionalWriteable(SourceConfig::new);
+        aggregationConfig = in.readOptionalWriteable(AggregationConfig::new);
```

Comment:
It seems like there are unnecessarily many levels where `null` is allowed.  You're allowing `aggregationConfig` to be `null` here, but also in `AggregationConfig` `aggregatorFactoryBuilder` is allowed to be `null`.  I think at most one of these possibilities should be allowed.

Reply:
fixed in `AggregationConfig` - not aiming for at the moment but in future I see the possibility of this builder to not use aggregations.